internal/controllers/state/statusupdater.go

```diff
@@ -31,40 +53,71 @@ func NewStatusUpdater(kubeClient clients.ClientsInterface) StatusUpdater {
+		meta.SetStatusCondition(&o.Status.Conditions, metav1.Condition{Type: v1beta1.SpecialResourceReady, Status: metav1.ConditionFalse, Reason: Progressing})
+		meta.SetStatusCondition(&o.Status.Conditions, metav1.Condition{Type: v1beta1.SpecialResourceErrored, Status: metav1.ConditionFalse, Reason: Progressing})
+	})
+}
+
+// SetAsProgressing changes SpecialResource's Ready condition as true and changes Progressing and Errored conditions to false, and updates the status in the API.
+func (su *statusUpdater) SetAsReady(ctx context.Context, sr *v1beta1.SpecialResource, reason, message string) error {
+	return su.updateWithMutator(ctx, sr, func(o *v1beta1.SpecialResource) {
+		meta.SetStatusCondition(&o.Status.Conditions, metav1.Condition{Type: v1beta1.SpecialResourceReady, Status: metav1.ConditionTrue, Reason: reason, Message: message})
+
+		meta.SetStatusCondition(&o.Status.Conditions, metav1.Condition{Type: v1beta1.SpecialResourceProgressing, Status: metav1.ConditionFalse, Reason: Ready})
+		meta.SetStatusCondition(&o.Status.Conditions, metav1.Condition{Type: v1beta1.SpecialResourceErrored, Status: metav1.ConditionFalse, Reason: Ready})
+	})
+}
+
+// SetAsErrored changes SpecialResource's Errored condition as true and changes Ready and Progressing conditions to false, and updates the status in the API.
+func (su *statusUpdater) SetAsErrored(ctx context.Context, sr *v1beta1.SpecialResource, reason, message string) error {
+	return su.updateWithMutator(ctx, sr, func(o *v1beta1.SpecialResource) {
+		meta.SetStatusCondition(&o.Status.Conditions, metav1.Condition{Type: v1beta1.SpecialResourceErrored, Status: metav1.ConditionTrue, Reason: reason, Message: message})
+
+		meta.SetStatusCondition(&o.Status.Conditions, metav1.Condition{Type: v1beta1.SpecialResourceReady, Status: metav1.ConditionFalse, Reason: Errored})
+		meta.SetStatusCondition(&o.Status.Conditions, metav1.Condition{Type: v1beta1.SpecialResourceProgressing, Status: metav1.ConditionFalse, Reason: Errored})
+	})
+}
+
+func (su *statusUpdater) updateWithMutator(ctx context.Context, sr *v1beta1.SpecialResource, mutator func(sr *v1beta1.SpecialResource)) error {
 
 	update := v1beta1.SpecialResource{}
 
 	// If we cannot find the SR than something bad is going on ..
-	objectKey := types.NamespacedName{Name: sr.GetName(), Namespace: sr.GetNamespace()}
-	err := su.kubeClient.Get(ctx, objectKey, &update)
+	err := su.kubeClient.Get(ctx, types.NamespacedName{Name: sr.GetName(), Namespace: sr.GetNamespace()}, &update)
 	if err != nil {
-		utils.WarnOnError(errors.Wrap(err, "Is SR being deleted? Cannot get current instance"))
-		return
+		return errors.Wrap(err, "Is SR being deleted? Cannot get current instance")
 	}
 
-	update.Status.State = state
+	if sr.Status.Conditions == nil {
+		sr.Status.Conditions = make([]metav1.Condition, 0)
+	}
+
+	mutator(&update)
 	update.DeepCopyInto(sr)
```

Comment:
I'm not sure why we copy the object here. Any idea? Could we remove that convoluted logic?

Reply:
i agree with @qbarrand , i think that those functions should be more straightforward:
1) no need to check for CR presence - in case it is missing, update will fail and we go for another loop where we will check the presence of the CR
2) SetAs.. function should return error (maybe not SetAsError) and reconciliation loop should fail in case CR is not updated
3) no need for deep copy, we can work on the current CR

Reply:
Thanks for the comments, I'll check it out

Reply:
btw. I think that the flow here (Get + Update) is defensive in case of CR change in API. I'll watch out for this when investigating simplification of this function

Reply:
Made changes as suggested. Had to change some `SpecialResource` params or struct fields to pointers.